### PR TITLE
Fix solution reload and config change issues

### PR DESCRIPTION
**Pull request by dsyme**
This fixes a bunch of bad regressions around solution reload and "change configuration" actions

* [All features stop working on any file after solution is reloaded](#2943)
* [All features stop working on opened script files after project reloading](#2916)
* [Changes in referenced projects are not seen until restart](#2872)

The problems were:

- we were caching the project options in LanguageService.fs but not invalidating the cache when a project was updated
- we were caching colorization information but not making that conditional on the defines active for a colorization
- we were not noticing "global" updates to project configuration affecting all projects, so getting situations where configurations were mixed between projects

Testing:
* [x] I've added a project under tests\projects\misc for manually checking behaviour when switching configs or making other changes to a project
* [x] I've checked the following load 

    devenv tests\projects\stress\big\dense\Dense.sln
    devenv tests\projects\stress\huge\dense\Dense.sln
    devenv tests\projects\stress\big\shallow\Shallow.sln

Notes on testing:
* The unit testing for much of this "the project is changed in Visual Studio" area has effectively been disabled since the Roslynization of the codebase, which we're certainly discovered by having lots of bugs in this area.  We will somehow have to rewrite or rejig unit tests.  For example, [this unit test](https://github.com/Microsoft/visualfsharp/blob/baa98f16e0b1a650f926f73512d1a1b6139ec094/vsintegration/tests/unittests/Tests.LanguageService.Completion.fs#L4315) is in theory "active" but it still tests only the unit-testable slice of the old FSharp.LanguageService.dll instead of FSharp.Editor.dll.  This means that it test how FSharp.LanguageService.dll is listening in and reacting to events from VS and FSharp.ProjectSystem.dll.  It doesn't test any of the code in FSharpEditor\...\LanguageService.fs 
* We should probably accept this PR and any other fixes to regressions before enabling the unit tests, but should activate the unit tests (somehow - it's really not easy!) before adding significant complexity to FSharp.Editor.  It would be great if we could activate the unit tests before replacing the F# Project System but maybe that's just going to make them all obsolete in any case....

@saul @vasily-kirichenko and others - I'd be grateful if you could try this out and validate that it fixes the bugs I've marked. Also check for loss of performance - we get a lot of "Project has updated" signals so have to recompute a lot of flags.

### CODE REVIEW

File: src/fsharp/CompileOps.fs [hunk 1]
```diff
@@ -2702,8 +2702,6 @@ type TcConfig private (data : TcConfigBuilder,validate:bool) =
         | Some(fslibFilename) ->
             let filename = ComputeMakePathAbsolute data.implicitIncludeDir fslibFilename
             try 
-                use ilReader = OpenILBinary(filename,data.optimizeForMemory,data.openBinariesInMemory,None,None, data.shadowCopyReferences)
```
Comment (dsyme) [inline src/fsharp/CompileOps.fs]: We check this again later, see other calls to ``checkFSharpBinaryCompatWithMscorlib``.  there is no need to do it here - opening the binary fails in Visual Studio if the binary doesn't exist on disk

File: src/fsharp/InternalCollections.fs [hunk 2]
```diff
@@ -13,7 +13,7 @@ type internal ValueStrength<'T when 'T : not struct> =
    | Weak of WeakReference<'T>
 #endif
 
-type internal AgedLookup<'Token, 'Key, 'Value when 'Value : not struct>(keepStrongly:int, areSame, ?requiredToKeep, ?onStrongDiscard, ?keepMax: int) =
+type internal AgedLookup<'Token, 'Key, 'Value when 'Value : not struct>(keepStrongly:int, areSimilar, ?requiredToKeep, ?onStrongDiscard, ?keepMax: int) =
```
Comment (dsyme) [inline src/fsharp/InternalCollections.fs]: This is just renaming arguments for consistency

File: src/fsharp/vs/service.fs [hunk 3]
```diff
@@ -2589,31 +2589,37 @@ type BackgroundCompiler(referenceResolver, projectCacheSize, keepAssemblyContent
           })
             
     member bc.InvalidateConfiguration(options : FSharpProjectOptions) =
+        // This operation can't currently be cancelled nor awaited
```
Comment (dsyme) [inline src/fsharp/vs/service.fs]: Invalidate configuration is adjusted to explicitly invalidate any "related" project

File: src/fsharp/vs/service.fsi [hunk 4]
```diff
@@ -401,7 +410,7 @@ type internal FSharpChecker =
     ///     can be used to marginally increase accuracy of intellisense results in some situations.
     /// </param>
     ///
-    member CheckFileInProjectIfReady : parsed: FSharpParseFileResults * filename: string * fileversion: int * source: string * options: FSharpProjectOptions * ?textSnapshotInfo: obj -> Async<FSharpCheckFileAnswer option>
+    member CheckFileInProjectAllowingStaleCachedResults : parsed: FSharpParseFileResults * filename: string * fileversion: int * source: string * options: FSharpProjectOptions * ?textSnapshotInfo: obj -> Async<FSharpCheckFileAnswer option>
```
Comment (dsyme) [inline src/fsharp/vs/service.fsi]: renaming

File: vsintegration/src/FSharp.Editor/LanguageService/LanguageService.fs [hunk 5]
```diff
@@ -64,6 +64,8 @@ type internal FSharpCheckerProvider
 
     member this.Checker = checker.Value
 
+type Refreshable<'T> = 'T * (unit -> 'T)
```
Comment (dsyme) [inline vsintegration/src/FSharp.Editor/LanguageService/LanguageService.fs]: The idea here is that we have a function to recalc the options when we need to, But we also can deliver the latest copy of the options to other callers

File: vsintegration/src/FSharp.Editor/LanguageService/Tokenizer.fs [hunk 6]
```diff
@@ -267,7 +268,7 @@ module internal Tokenizer =
                 data.[i] <- None
                 i <- i + 1
 
-    let private dataCache = ConditionalWeakTable<DocumentId, SourceTextData>()
+    let private dataCache = ConditionalWeakTable<DocumentId, ConcurrentDictionary<string list, SourceTextData>>()
```
Comment (dsyme) [inline vsintegration/src/FSharp.Editor/LanguageService/Tokenizer.fs]: This is a fix to colorization - when switching configurations, colorization information was not updated, e.g. ``#if DEBUG`` sections.

**Comment by dsyme** [comment]
Some glitches still need to be fixed

* [ ] Still getting  ``...\Debug\net40\bin\FSharp.Core.dll' was not found or is invalid``` from VisualFSharp.sln.  We should not be getting that when in Release configuration.  This error went away for smaller solutions.

* [ ] Two failures due to loading FSharp.Core.dll from disk

* [ ] I tried to re-validate that ``#if`` colorization has the right behaviour on configuration change but I haven't been able to do that yet

```
+++ MultiTargeting (E_BadPathToFSharpCore.fs) +++
--------------------------------------------------------
Error from hosted compiler
Exit code: 1
Error:     2
Microsoft (R) F# Compiler version 4.1
Copyright (c) Microsoft Corporation. All Rights Reserved.
(0,0): error FS0082: Could not resolve this reference. Could not locate the assembly "FSharp.Core.dll". Check to make sure the assembly exists on disk. If this reference is required by your code, you may get compilation errors. (Code=MSB3245)
(0,0): error FS0073: internal error: BuildFrameworkTcImports: no resolution of 'FSharp.Core.dll'
--------------------------------------------------------
Microsoft (R) F# Compiler version 4.1
Copyright (c) Microsoft Corporation. All Rights Reserved.
(0,0): error FS0082: Could not resolve this reference. Could not locate the assembly "FSharp.Core.dll". Check to make sure the assembly exists on disk. If this reference is required by your code, you may get compilation errors. (Code=MSB3245)
(0,0): error FS0073: internal error: BuildFrameworkTcImports: no resolution of 'FSharp.Core.dll'

*** The following necessary lines were never matched:
***	error FS0229:.+Error opening binary file '.+I_DO_NOT_EXIST\\FSharp\.Core\.dll'


*** The following necessary lines were incorrectly matched:
***	(0,0): error FS0073: internal error: BuildFrameworkTcImports: no resolution of 'FSharp.Core.dll'


Unexpected Compiler Output 
FAIL
+++ MultiTargeting (E_BadPathToFSharpCore.fsx) +++
--------------------------------------------------------
Error from hosted compiler
Exit code: 1
Error:     2
Microsoft (R) F# Compiler version 4.1
Copyright (c) Microsoft Corporation. All Rights Reserved.
(0,0): error FS0082: Could not resolve this reference. Could not locate the assembly "FSharp.Core.dll". Check to make sure the assembly exists on disk. If this reference is required by your code, you may get compilation errors. (Code=MSB3245)
--------------------------------------------------------
Microsoft (R) F# Compiler version 4.1
Copyright (c) Microsoft Corporation. All Rights Reserved.
(0,0): error FS0082: Could not resolve this reference. Could not locate the assembly "FSharp.Core.dll". Check to make sure the assembly exists on disk. If this reference is required by your code, you may get compilation errors. (Code=MSB3245)

*** The following necessary lines were never matched:
***	error FS0229:.+Error opening binary file '.+I_DO_NOT_EXIST\\FSharp\.Core\.dll'


*** The following necessary lines were incorrectly matched:

Unexpected Compiler Output 
FAIL
```

**Comment by vasily-kirichenko** [comment]
1. Load VFT solution
2. Change the solution file outside VS
3. Reload the solution
4. Everything works, coloring reappear much quicker than the first time
5. The following assertion raised (it seems to be unrelated to this PR though)

```
---------------------------
Assertion Failed: Abort=Quit, Retry=Debug, Ignore=Continue
---------------------------
IncrementalBuild object has already been disposed!



   at Microsoft.FSharp.Compiler.IncrementalBuilder.assertNotDisposed()

   at Microsoft.FSharp.Compiler.IncrementalBuilder.DecrementUsageCount()

   at <StartupCode$FSharp-LanguageService-Compiler>.$Service.-ctor@2157-123.Invoke(Tuple`3 tupledArg)

   at Internal.Utilities.Collections.AgedLookup`3.strongDiscard(Value x)

   at <StartupCode$FSharp-LanguageService-Compiler>.$InternalCollections.clo@110-9.Invoke(Value x)

   at Microsoft.FSharp.Primitives.Basics.List.iter[T](FSharpFunc`2 f, FSharpList`1 x)

   at Microsoft.FSharp.Collections.ListModule.Iterate[T](FSharpFunc`2 action, FSharpList`1 list)

   at Internal.Utilities.Collections.AgedLookup`3.AssignWithStrength[d,e](d tok, FSharpList`1 newdata, FSharpList`1 discard1)

   at <StartupCode$FSharp-LanguageService-Compiler>.$Service.clo@2166-424.Invoke(Tuple`3 _arg3)

   at Microsoft.FSharp.Compiler.AbstractIL.Internal.Library.CancellableModule.bind@744.Invoke(CancellationToken ct)

   at Microsoft.FS......

<truncated>
---------------------------
Abort   Retry   Ignore   
---------------------------
```

**Comment by dsyme** [comment]
This is ready (once green)

**Comment by dsyme** [comment]
@vasily-kirichenko If you could give it another go that would be great.

**Comment by dsyme** [comment]
@vasily-kirichenko @KevinRansom @brettfo @saul and others - could I request a code review on this please?  Thanks

File: src/fsharp/InternalCollections.fsi [hunk 7]
```diff
@@ -44,21 +44,30 @@ namespace Internal.Utilities.Collections
     new : keepStrongly:int 
             * areSame:('Key * 'Key -> bool) 
             * ?isStillValid:('Key * 'Value -> bool)
-            * ?areSameForSubsumption:('Key * 'Key -> bool) 
+            * ?areSimilar:('Key * 'Key -> bool)
```
Comment (vasily-kirichenko) [inline src/fsharp/InternalCollections.fsi]: A XML Doc comment would be great here explaining what "similarity" means here and what it will be used for.

File: vsintegration/src/FSharp.Editor/LanguageService/LanguageService.fs [hunk 8]
```diff
@@ -73,30 +76,47 @@ type internal ProjectInfoManager
...
-    member this.RemoveSingleFileProject(projectId) =
+    member this.ClearInfoForSingleFileProject(projectId) =
         singleFileProjectTable.TryRemove(projectId) |> ignore
 
-    /// Clear a project from the project table
-    member this.ClearProjectInfo(projectId: ProjectId) =
-        projectTable.TryRemove(projectId) |> ignore
+    member this.RefreshInfoForProjectsThatReferenceThisProject(projectId: ProjectId) =
+        // Search the projectTable for things to refresh
+        for KeyValue(otherProjectId, ((referencedProjectIds, _options), refresh)) in projectTable.ToArray() do
```
Comment (vasily-kirichenko) [inline vsintegration/src/FSharp.Editor/LanguageService/LanguageService.fs]: I'm just curious, why `ToArray` here?

**Comment by vasily-kirichenko** [comment]
> @dsyme If you could give it another go that would be great.

Same endless assertions.

**Comment by majocha** [comment]
I can confirm the 3 issues fixed.
When switching branches in VisualFSharp I got lots of identical:
`[Failure] Could not find file 'C:\Users\Kuba\AppData\Local\Temp\.NETFramework,Version=v4.5.AssemblyAttributes.cs'.`
But I think I've seen those before.

File: src/fsharp/CompileOps.fs [hunk 9]
```diff
@@ -2701,13 +2701,15 @@ type TcConfig private (data : TcConfigBuilder,validate:bool) =
...
         | Some(fslibFilename) ->
             let filename = ComputeMakePathAbsolute data.implicitIncludeDir fslibFilename
-            try 
-                use ilReader = OpenILBinary(filename,data.optimizeForMemory,data.openBinariesInMemory,None,None, data.shadowCopyReferences)
-                checkFSharpBinaryCompatWithMscorlib filename ilReader.ILAssemblyRefs ilReader.ILModuleDef.ManifestOfAssembly.Version rangeStartup;
-                let fslibRoot = Path.GetDirectoryName(FileSystem.GetFullPathShim(filename))
-                fslibRoot (* , sprintf "v%d.%d" v1 v2 *)
-            with e -> 
-                error(Error(FSComp.SR.buildErrorOpeningBinaryFile(filename, e.Message), rangeStartup))
+            if (* validate && *) fslibReference.ProjectReference.IsNone then
```
Comment (saul) [inline src/fsharp/CompileOps.fs]: Why validate commented?

File: vsintegration/src/FSharp.Editor/LanguageService/LanguageService.fs [hunk 10]
```diff
@@ -64,6 +65,8 @@ type internal FSharpCheckerProvider
 
     member this.Checker = checker.Value
 
+type Refreshable<'T> = 'T * (bool -> 'T)
```
Comment (saul) [inline vsintegration/src/FSharp.Editor/LanguageService/LanguageService.fs]: XML doc comment here?

File: vsintegration/src/FSharp.Editor/LanguageService/LanguageService.fs [hunk 11]
```diff
@@ -308,39 +347,58 @@ and
...
-        for ref in workspaceRefs do
+        for ref in originalRefs do
             if not(updatedRefs.Contains(ref)) then
                 projectContext.RemoveMetadataReference(ref)
                 updated <- true
 
+        let ok,originalOptions = optionsAssociation.TryGetValue(projectContext)
+        let updatedOptions = site.CompilerFlags()
+        if not ok || originalOptions <> updatedOptions then 
+            let updatedFiles = site.SourceFilesOnDisk() |> wellFormedFilePathSetIgnoreCase
```
Comment (saul) [inline vsintegration/src/FSharp.Editor/LanguageService/LanguageService.fs]: Shadowing the exact same above

File: vsintegration/src/FSharp.Editor/LanguageService/LanguageService.fs [hunk 12]
```diff
@@ -308,39 +347,58 @@ and
...
                 projectContext.AddMetadataReference(ref, MetadataReferenceProperties.Assembly)
                 updated <- true
 
-        for ref in workspaceRefs do
+        for ref in originalRefs do
             if not(updatedRefs.Contains(ref)) then
                 projectContext.RemoveMetadataReference(ref)
                 updated <- true
 
+        let ok,originalOptions = optionsAssociation.TryGetValue(projectContext)
```
Comment (saul) [inline vsintegration/src/FSharp.Editor/LanguageService/LanguageService.fs]: match expression may be more succinct

File: vsintegration/src/FSharp.LanguageService/ProjectSitesAndFiles.fs [hunk 13]
```diff
@@ -110,45 +110,49 @@ type internal ProjectSitesAndFiles() =
...
-                    // In this case we go and compute the options recursively.
-                    match tryGetOptionsForReferencedProject p.FileName with 
-                    | None -> getProjectOptionsForProjectSite (tryGetOptionsForReferencedProject, ps.GetProjectSite(), fileName, extraProjectInfo, serviceProvider, useUniqueStamp)
-                    | Some options -> options
-                path, referencedProjectOptions)
-            )
-        |> Seq.toArray
+    static let rec referencedProjectsOf (tryGetOptionsForReferencedProject, projectSite:IProjectSite, extraProjectInfo, serviceProvider:System.IServiceProvider, useUniqueStamp) =
+        [| for (p,ps) in referencedProvideProjectSites (projectSite, serviceProvider) do
+               match fullOutputAssemblyPath p with
```
Comment (saul) [inline vsintegration/src/FSharp.LanguageService/ProjectSitesAndFiles.fs]: is the indentation off by one here?

File: vsintegration/src/FSharp.ProjectSystem.FSharp/Project.fs [hunk 14]
```diff
@@ -1739,49 +1726,66 @@ namespace rec Microsoft.VisualStudio.FSharp.ProjectSystem
...
+                    VSConstants.S_OK
                 member x.UpdateProjectCfg_Begin(pHierProj, _pCfgProj, _pCfgSln, _dwAction, pfCancel) =
                     UpdateConfig(pHierProj)
                     VSConstants.S_OK
                 member x.UpdateProjectCfg_Done(pHierProj, _pCfgProj, _pCfgSln, _dwAction, _fSuccess, _fCancel) =
                     UpdateConfig(pHierProj)
+                    // We wait until allproject configurations have been updated before updating our MSBuild state
+                    projNode.UpdateMSBuildState()
+                    projNode.SetProjectFileDirty(projNode.IsProjectFileDirty)
+                    projNode.ComputeSourcesAndFlags()
```
Comment (saul) [inline vsintegration/src/FSharp.ProjectSystem.FSharp/Project.fs]: When is UpdateProjectCfg_Done called when OnAfterActiveSolutionCfgChange isn't? Are we definitely only calling ComputeSourcesAndFlags once per project on configuration change? Also is the wait dialog still open when we call ComputeSourcesAndFlags here (is this method called before OnAfterActiveSolutionCfgChange)?

**Comment by dsyme** [comment]
@saul Thanks for the review!  I'll go through the items now

File: vsintegration/src/FSharp.ProjectSystem.FSharp/Project.fs (hunk 14 above)
Comment (dsyme) [inline vsintegration/src/FSharp.ProjectSystem.FSharp/Project.fs]: Here's what I see for the set of events triggered when loading ``tests\projects\misc\TestProjectChanges.sln``. Looking at this, 


* I believe you are right that we are doing ComputeSourcesAndFlags twice per project but only because we are getting two complete sets of ``OnActiveProjectCfgChangeBatchBegin``/ ``OnActiveProjectCfgChangeBatchEnd`` events sent through.

* It is important not to do any ComputeSourcesAndFlags until we have updated the configuration on all the projects. So I think we have to do this in OnActiveProjectCfgChangeBatchEnd.

* The cost of ``ComputeSourcesAndFlags`` has been much reduced since #3002 since it uses the stored computed options for referenced projects


### Loading tests\projects\misc\TestProjectChanges.sln

* OnActiveProjectCfgChangeBatchBegin x 3
* IVsUpdateSolutionEvents.OnActiveProjectCfgChange(null) x 3
* IVsUpdateSolutionEvents.OnActiveProjectCfgChange(not-null) x 6
* OnActiveProjectCfgChangeBatchEnd x 3

Then we get a "duplicate" set of Batch events
* OnActiveProjectCfgChangeBatchBegin x 3
* OnActiveProjectCfgChangeBatchEnd x 3

### Switching to "Release"
* OnBeforeActiveSolutionCfgChange x 3
* OnActiveProjectCfgChangeBatchBegin x 3
* IVsUpdateSolutionEvents.OnActiveProjectCfgChange(null) x 3
* IVsUpdateSolutionEvents.OnActiveProjectCfgChange(not-null) x 6
* OnActiveProjectCfgChangeBatchEnd x 3
* OnAfterActiveSolutionCfgChange x 3

### On prompted solution reload after a project file has been edited
* OnActiveProjectCfgChangeBatchBegin x 3
* IVsUpdateSolutionEvents.OnActiveProjectCfgChange(null) x 3
* IVsUpdateSolutionEvents.OnActiveProjectCfgChange(not-null) x 6
* OnActiveProjectCfgChangeBatchEnd x 3

Then we get a "duplicate" set of Batch events

* OnActiveProjectCfgChangeBatchBegin x 3
* OnActiveProjectCfgChangeBatchEnd x 3

### On individual project reload 

* OnActiveProjectCfgChange 

Also I never see these being called:
* IVsUpdateSolutionEvents2.OnActiveProjectCfgChange
* IVsUpdateSolutionEvents2.UpdateProjectCfg_Begin
* IVsUpdateSolutionEvents2.UpdateProjectCfg_Done

File: src/fsharp/CompileOps.fs (hunk 9 above)
Comment (dsyme) [inline src/fsharp/CompileOps.fs]: I've removed the commented out text now.  It turns out this must be being called by ``fsc.exe`` with ``validate=false``, so the error message sequence was being changed by checking this, but that is a separate issue

File: vsintegration/src/FSharp.Editor/LanguageService/LanguageService.fs (hunk 11 above)
Comment (dsyme) [inline vsintegration/src/FSharp.Editor/LanguageService/LanguageService.fs]: Thanks, good catch, fixed

File: vsintegration/src/FSharp.Editor/LanguageService/LanguageService.fs (hunk 10 above)
Comment (dsyme) [inline vsintegration/src/FSharp.Editor/LanguageService/LanguageService.fs]: thanks

File: vsintegration/src/FSharp.LanguageService/ProjectSitesAndFiles.fs (hunk 13 above)
Comment (dsyme) [inline vsintegration/src/FSharp.LanguageService/ProjectSitesAndFiles.fs]: thanks

**Comment by dsyme** [comment]
@saull Could you check the last two commits?  I've updated to only call ``ComputeSourcesAndFlags`` in ``OnAfterActiveSolutionCfgChange`` and this seems to have the desired effect of only doing the operation once on load, reload and configuration update.  On load and reload other operations end up forcing ComputeSourcesAndFlags

After this change I don't see any value in the ``batchState`` tracking.

I must say this logic is a right mess.  I can't wait to ditch the whole project system and replace it with the Roslyn one

**Comment by dsyme** [comment]
@saul OK, I think this is ready

@vasily-kirichenko I believe the assertion is a different problem - we've certainly seen it before - I've been able to repro it sometimes but not consistently.

**Comment by dsyme** [comment]
> When switching branches in VisualFSharp I got lots of identical: [Failure] Could not find file 'C:\Users\Kuba\AppData\Local\Temp\.NETFramework,Version=v4.5.AssemblyAttributes.cs'.

I've not seen this before, I'm confident it's a separate problem but it would be good to address it

File: vsintegration/src/FSharp.ProjectSystem.FSharp/Project.fs [hunk 15]
```diff
@@ -1674,114 +1672,167 @@ namespace rec Microsoft.VisualStudio.FSharp.ProjectSystem
...
-                    if batchState = BatchWaiting then
-                        batchState <- BatchDone
-                VSConstants.S_OK
+                        null 
 
             let UpdateConfig(pHierProj) =
-                // By default, the F# project system keeps its own internal Configuration and Platform in sync with the current active
-                // Configuration and Platform by listening for OnActiveProjectCfgChange events.  However there is one case where the
-                // active cfg changes without an event, and this is during 'Batch Build'.  So we listen for the start and end of 
-                // Batch Build, and manually update the project to the active cfg before/after to set/reset the config.
```
Comment (saul) [inline vsintegration/src/FSharp.ProjectSystem.FSharp/Project.fs]: Why has this comment been removed? Is this no longer the case? Your other comment didn't lead me to believe that you tried doing a batch build.

Comment (saul) [inline vsintegration/src/FSharp.ProjectSystem.FSharp/Project.fs]: Why has this comment been removed? Is this no longer the case? Your other comment didn't lead me to believe that you tried doing a batch build.

File: vsintegration/src/FSharp.ProjectSystem.FSharp/Project.fs [hunk 16]
```diff
@@ -1674,114 +1672,167 @@ namespace rec Microsoft.VisualStudio.FSharp.ProjectSystem
...
+                            }
+                            |> WaitDialog.start projNode.Site
+                            |> Some
 
                     VSConstants.S_OK
 
                 member x.OnAfterActiveSolutionCfgChange(_oldCfg, _newCfg) =
+
+                    // We have updated the MSBuild state of each project.  Now we can call ComputeSourcesAndFlags
+                    projNode.ComputeSourcesAndFlags()
```
Comment (saul) [inline vsintegration/src/FSharp.ProjectSystem.FSharp/Project.fs]: Why are we doing this here? This means that the progress dialog is going to close way before all of the ComputeSourcesAndFlags are complete.

We should reference count how many times waitDialog was (going to be) opened, and then only dispose of it when it reaches 0.

File: vsintegration/src/FSharp.ProjectSystem.FSharp/Project.fs [hunk 17]
```diff
@@ -1674,114 +1672,167 @@ namespace rec Microsoft.VisualStudio.FSharp.ProjectSystem
...
                     VSConstants.S_OK
               
             interface IVsUpdateSolutionEvents4 with
                 member x.OnActiveProjectCfgChangeBatchBegin() =
-                    batchState <- BatchWaiting
+                    ()
+
                 member x.OnActiveProjectCfgChangeBatchEnd() =
-                    batchState <- NonBatch
+                    projNode.UpdateMSBuildState()
```
Comment (saul) [inline vsintegration/src/FSharp.ProjectSystem.FSharp/Project.fs]: Any comments as to why this is now necessary? Why at this stage and not at OnActiveProjectCfgChange?

File: vsintegration/src/FSharp.ProjectSystem.FSharp/Project.fs [hunk 18]
```diff
@@ -1674,114 +1672,167 @@ namespace rec Microsoft.VisualStudio.FSharp.ProjectSystem
...
-                // Configuration and Platform by listening for OnActiveProjectCfgChange events.  However there is one case where the
-                // active cfg changes without an event, and this is during 'Batch Build'.  So we listen for the start and end of 
-                // Batch Build, and manually update the project to the active cfg before/after to set/reset the config.
                 if GetCaption(pHierProj) = GetCaption(projNode.InteropSafeIVsHierarchy) then
                     // This code matches what ProjectNode.SetConfiguration would do; that method cannot be called during a build, but at this
                     // current moment in time, it is 'safe' to do this update.
                     let _,currentConfigName = Utilities.TryGetActiveConfigurationAndPlatform(projNode.Site, projNode.ProjectIDGuid)
                     MSBuildProject.SetGlobalProperty(projNode.BuildProject, ProjectFileConstants.Configuration, currentConfigName.ConfigName)
                     MSBuildProject.SetGlobalProperty(projNode.BuildProject, ProjectFileConstants.Platform, currentConfigName.MSBuildPlatform)
-                    projNode.UpdateMSBuildState()
```
Comment (saul) [inline vsintegration/src/FSharp.ProjectSystem.FSharp/Project.fs]: Why was this removed?

File: vsintegration/src/FSharp.ProjectSystem.FSharp/Project.fs (hunk 17 above)
Comment (dsyme) [inline vsintegration/src/FSharp.ProjectSystem.FSharp/Project.fs]: All good questions.  TBH I just fiddled around for about 3 days (yes really) until it seemed to work in every scenario I tried it in.  Almost every other variation seemed to fail in some way or another.

I didn't however try a batch build.  I had misunderstood what that actually was, sorry.

You are correct that the dialog goes away early though another "computing sources and flags for project" replaces it.  

I'll see if I can find a less invasive set of changes.  Perhaps there is just one fix needed in this code after all.

**Comment by dsyme** [comment]
@KevinRansom This is _not_ yet ready to go in (at least until @saul approves :))

**Comment by dsyme** [comment]
Here's what I see for the set of events triggered when loading ``tests\projects\misc\TestProjectChanges.sln``. Looking at this, 


* I believe you are right that we are doing ComputeSourcesAndFlags twice per project on solution load.  But suspect we were doing that before, and it is because we are getting two complete sets of ``OnActiveProjectCfgChangeBatchBegin``/ ``OnActiveProjectCfgChangeBatchEnd`` events sent through.

* The cost of ``ComputeSourcesAndFlags`` has been much reduced since #3002 since it uses the stored computed options for referenced projects

* It is important not to do any ComputeSourcesAndFlags until we have updated the configuration on all the projects. So I think we have to do this in OnActiveProjectCfgChangeBatchEnd.


### Loading tests\projects\misc\TestProjectChanges.sln

* OnActiveProjectCfgChangeBatchBegin x 3
* IVsUpdateSolutionEvents.OnActiveProjectCfgChange(null) x 3
* IVsUpdateSolutionEvents.OnActiveProjectCfgChange(not-null) x 6
* OnActiveProjectCfgChangeBatchEnd x 3 

Then we get a "duplicate" set of Batch events
* OnActiveProjectCfgChangeBatchBegin x 3
* OnActiveProjectCfgChangeBatchEnd x 3 

Then something else forces **ComputeSourcesAndFlags** on each project

### Switching to "Release"
* OnBeforeActiveSolutionCfgChange x 3
* OnActiveProjectCfgChangeBatchBegin x 3
* IVsUpdateSolutionEvents.OnActiveProjectCfgChange(null) x 3
* IVsUpdateSolutionEvents.OnActiveProjectCfgChange(not-null) x 6
* OnActiveProjectCfgChangeBatchEnd x 3 
* OnAfterActiveSolutionCfgChange x 3 **ComputeSourcesAndFlags** 

### On prompted solution reload after a project file has been edited
* OnActiveProjectCfgChangeBatchBegin x 3
* IVsUpdateSolutionEvents.OnActiveProjectCfgChange(null) x 3
* IVsUpdateSolutionEvents.OnActiveProjectCfgChange(not-null) x 6
* OnActiveProjectCfgChangeBatchEnd x 3 

Then we get a "duplicate" set of Batch events

* OnActiveProjectCfgChangeBatchBegin x 3
* OnActiveProjectCfgChangeBatchEnd x 3 

Then something else forces **ComputeSourcesAndFlags** on each project

### On individual project reload 

* OnActiveProjectCfgChange 

### On batch build:

* UpdateProjectCfg_Begin x 6 (twice for each project!)
* UpdateProjectCfg_Done x 6 (twice for each project!)


Also I never see these being called in the scenarios I've tested - if you know a sequence that triggers them please let me know :)
* IVsUpdateSolutionEvents2.OnActiveProjectCfgChange
* IVsUpdateSolutionEvents2.UpdateProjectCfg_Begin
* IVsUpdateSolutionEvents2.UpdateProjectCfg_Done

**Comment by dsyme** [comment]
@saul Here are my further investigations

- First, I have adjusted the code to keep a count for the dialog

- Second, I have tested "Batch Build".  This functions correctly

- Third, I've put the ``batchState`` value back in.  This actually has nothing to do with batch builds - the ``OnActiveProjectCfgChangeBatchBegin`` doesn't get invoked for batch build and I really do think the old comments were just misled (a lot of this code is awful - we should never have copied the original C# VS sample from whence all this stems - and I can't wait to get rid of it with the new project system).  I've added an assert that ``batchState`` is always ``NonBatch`` by the time we get to ``OnAfterActiveSolutionCfgChange``

- Fourth, I tried to move ``ComputeSourcesAndFlags`` back to ``OnActiveProjectCfgChangeBatchEnd``.  The problem with moving it there or any earlier is that it seems the MSBuild state of the project is out of sync, and we get invalid mixes of Debug and Release references in project options, This is the primary problem I've been fighting and which ``TestProjectChanges`` helps test (maually).  I've tried hard to work out **why** this happens, and whether it is a regression or an existing bug, but it's just too hard for me.  So I've kept ``ComputeSourcesAndFlags`` in ``OnAfterActiveSolutionCfgChange`` and made sure the dialog stays up.

So in short we should have the same number of ``ComputeSourcesAndFlags`` calls - 1 per project - but they just come later in the event sequence, when all projects have had their MSBuild state updated and the project options drop out correctly. Any other ordering just doesn't seem to work for me.

Manual testing:

tests\proejcts\misc\TestProjectChanges\TestProjectChanges.sln
* Change to/from Debug/Release config and watch the intellisense/coloring update. No errors should appear
* Build in each configuration
* File --> Close Solution
* File --> Open Solution
* File --> Close Solution
* Do a batch build over all configurations  (except Lib2=Release which makes no sense for this)
* Do a batch build "clean" over all configurations  (except Lib2=Release which makes no sense for this)
* Do a batch rebuild over all configurations (except Lib2=Release which makes no sense for this)

VisualFSharp.sln
* Open in Debug, do a build, check ``#if DEBUG`` is active
* Switch to Release do a build, check ``#if DEBUG`` is NOT active
* File --> Close Solution
* File --> Open Solution
* File --> Close Solution

I think this is the best I can do.  Though I'll try again if you ask me :)

File: vsintegration/src/FSharp.ProjectSystem.FSharp/Project.fs (hunk 15 above)
Comment (dsyme) [inline vsintegration/src/FSharp.ProjectSystem.FSharp/Project.fs]: See comment below.

**Comment by dsyme** [comment]
@saul The thanks is to you for reviewing so carefully.  It's about the most important thing we can do.

**Comment by dsyme** [comment]
@KevinRansom this is ready to merge (unless you wish to add additional reviews)
